### PR TITLE
[NETBEANS-1438] : stackoverflowerror opening .js files

### PR DESCRIPTION
**Pull request by rtaneja1**
Added checks to prevent StackOverflowError on opening some .js files.

### CODE REVIEW

**Comment by Chris2011** [comment]
Is this really a solution to prevent a SOE? I mean, why does this happen?

**Comment by rtaneja1** [comment]
I agree, but this exception has been reported multiple times, without providing the exact .js code snippet that is causing issue.  The code causing the SOE was added for goto support which IMO is nice to have.

**Comment by Chris2011** [comment]
I think this PR needs a really deep look whether this will not causing other side effects.

**Comment by junichi11** [comment]
Thanks a lot for working on this, Reema!
I think that we should use this solution to just leave a log. Otherwise, it means that we just ignore the bug. I don't think it is good...

I guess that a reporter could not provide an example code because the reporter could not find a file which has a problem. So it is significant to leave the log, I think.

BTW, if the static field is accessed from multiple threads, it can be an incorrect value? Instead, we should use a method parameter? Am I wrong?

File: webcommon/javascript2.model/src/org/netbeans/modules/javascript2/model/api/ModelUtils.java [hunk 1]
```diff
@@ -1173,75 +1177,84 @@ public static boolean hasDeclaredProperty(JsObject jsObject) {
...
+                                object = ModelUtils.searchJsObjectByName(parent, newObjectName);
+                                for (JsObject prop : object.getProperties().values()) {
+                                    if ((prop.getName().equals(tokens[i])) && (prop.isDeclared())) {
+                                        if (prop.getAssignmentCount() > 0) {
+                                            if (invokeCount == MAX_RECURSION_DEEP_RESOLVING_ASSIGNMENTS) {
+                                                LOG.log(Level.WARNING, "StackOverFlowError : {0} : {1}",
+                                                        new Object[]{object.getFullyQualifiedName(),
+                                                            object.getFileObject()});
+                                            }
+                                            for (TypeUsage type1 : prop.getAssignments()) {
```
Comment (junichi11) [inline webcommon/javascript2.model/src/org/netbeans/modules/javascript2/model/api/ModelUtils.java]: Need this check? (Here is within the `if (prop.getAssignmentCount() > 0)` block and `type1` is not used.)
If need it, should write like this?
```java
if (!prop.getAssignments().isEmpty()) {
...
}
```

**Comment by junichi11** [comment]
Thank you for fixing it!
I've verified that SOE does not occur with this change :)

**Comment by junichi11** [comment]
@rtaneja1 Could you please squash before you merge this? Thanks.

**Comment by junichi11** [comment]
@rtaneja1 Please don't misunderstand. I'm not blaming you at all.

**Comment by rtaneja1** [comment]
Thanks for your help on this Junichi. Merging.

**Comment by junichi11** [comment]
@lkishalmi This issue is a blocker: https://issues.apache.org/jira/browse/NETBEANS-1438
So could you please cherry-pick this into the release branch if it's OK with you?

**Comment by lkishalmi** [comment]
Cherry picked into release100

**Pull request merged**